api/controllers/user-controller.js

```
@@ -0,0 +1,40 @@
+const { saveUser, getUsers } = require('../service/user-service');
+const { validationResult } = require('express-validator/check');
+const toValidationErrorString = require('../utilities/express-validator-helper');
+const { ok, fail } = require('../utilities/response-helper');
+
+const log = require('../utilities/logger');
+
+async function save(req, res, next) {
+  try {
```

Maybe move to line 16? I suggest to reduce the coverage of `try-catch`. We should only use it when something is really unexpectable. (e.g. DB connection here) Line 10 - 16 shouldn't be unexpectable.

---

I wouldn't exclude the validation code outside the try catch block. The strategy is to let any exception bubble up and captured at the highest layer which is the controller level. If there are any exception not being captured, it would simply crash the entire stacks and server would not return any response to client. The express-validation library is an abstraction layer to us where we don't have direct visibility of isEmpty method. We would need to make sure that it will not throw any exception, otherwise I think it is better to keep it inside the try catch.

---

We can definitely keep it. However, it's really an interesting topic which we can discuss more! 😁

My understanding is that `try-catch` can't really catch all errors happen inside. I know `await-async` can help, but it doesn't cover all use cases. Take the following as example, you can't catch the error inside the `setTimeout`.

http://jsbin.com/sesodit/edit?js,console
```
function foo() {
  try {
    setTimeout(function () {
      throw new Error('Catch me if you can');
    }, 1000);
  } catch (e) {
    console.log('Catched successfully - ' + e.message);
  }
}
foo();
```

We usually use a lot of third-party libraries as abstractions, especially in front-end. If the top-level `try-catch` couldn't catch all bubbled-up errors, does that mean we have to `try-catch` all of them? Should I use `try-catch` whenever I use abstractions like React and lodash which I don't have direct visibility? 

I am not familiar with express. My assumption that all errors will be caught by the default error handler like `window.onerror`, right?  If so, we should be able to avoid crash here? 

Let me know your thoughts :D

---

I think more deeply while I walk to the office. 

I sparely use `try-catch` in front-end for this project and company projects. I only use it when I want to bypass the error or print out the error specifically. I can keep the JavaScript working as long as I handle have `window.onerror` which returns true. 

For back-end, it's a different story because you can use `await/async`. If you don't handle it with `try-catch`, the error will be swallowed silently without throwing to express. The error by default is just a `console.error` message, not a real app error. It's overwhelming if I have to add `try-catch` for all third party libraries. 

Back to the `express-validator`, I think it's totally fine by not wrapping it in `try-catch` block. Express should be able to catch its error. I have checked the back-end implementation. You guys basically use `try-catch` to re-throw errors to express because `await/async` won't do that.

I should make a small express server to prove my thought. 😆

---

With setTimeout, it starts the work in a new thread. The try catch in the original thread is not going to capture any exception occurred in other threads. With async await (well, this is at least how it behaves in C#, I guess it can be somewhat similar in JS in a way), when the second thread completes the execution, the runtime restores the result (including the exception) back to the original thread, thus, if the original thread "awaits" the result, it would capture the exception that was thrown from the second thread. (However, if the original thread does not await for the result, it behaves just like the setTimeout function, the exception simply got swallowed).

My understanding of Express is that it is un-opinionated about the process and it doesn't handle any exceptions. What we have setup right now is that we have an error handler middleware that inspects all the outgoing response. When the controller catches any exception, it calls the next middleware and pass in the error. When the error handler middlware receives the error, it transforms the error into HTTP server error (500) and returns the response to client.

In our case, the controller is really our last defense line (just like your earlier comment about the browser). There might be a better way to handle exception but by default, Express does not handle uncaught exceptions.

---

@yhua045 

It's actually quite different with our guessing. I've written a prototype and have the following conclusion points. 

* Express CAN handle errors in synchronous code automatically by default. 
* Express CAN'T handle errors in asynchronous code automatically. Creator who makes the async function need to handle the errors accordingly. (e.g. For the Promise, `reject` error is required) User who use this method need to re-throw to with the `next` method.

According to this, wrapping a try-catch for a third-party library is not very useful. The error in synchronous method like express-validator's `validate` will be handled by Express automatically. And if it has asynchronous call inside which is not handled correctly, our App will still crash even you put try-catch. 

The following URLs all cause error without `try-catch` wrapping on the top-level except the `await/async` example. The server is still running good.

https://vast-sierra-93927.herokuapp.com/error
https://vast-sierra-93927.herokuapp.com/error-method
https://vast-sierra-93927.herokuapp.com/error-settimeout
https://vast-sierra-93927.herokuapp.com/error-async-await

```js
const express = require("express");
const app = express();
const fs = require("fs");

app.get("/", (req, res) => res.send("Hello World!"));

// Express will catch the errors in synchrous code #1
app.get("/error", (req, res) => {
  throw new Error("Trigger error #1"); // try-catch is not necessary
});

// Express will catch the errors in synchrous code #2
const foo = () => {
  throw new Error("foo is broken");
};
app.get("/error-method", () => {
  foo(); // try-catch is not necessary.
});

// Express can't handle error which happens in asynchrous code
// App breaks even if you place try-catch outside.
// The try-catch must happen inside.
// The error won't be swallowed.
app.get("/error-settimeout", (req, res, next) => {
  setTimeout(() => {
    try {
      throw new Error("Trigger error #2");
    } catch (e) {
      next(e);
    }
  }, 1000);
});

// For both .then and async/await syntax
// App will keep loading if we don't resolve nor reject.
const promiseFn = () => {
  return new Promise((resolve, reject) => {
    fs.readFile("/file-does-not-exist", "utf8", (err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(data);
      }
    });
  });
};

// The error will be swallowed if you don't use catch with next to re-throw the error
// The App will keep loading if you don't use catch with next to re-throw the error
app.get("/error-promise", (req, res, next) => {
  promiseFn()
    .then(data => {
      console.log(data);
    })
    .catch(e => {
      // It keeps loading if I don't specify catch.
      next("Catched promise error - " + e);
    });
});

// The error will be swallowed if you don't use try-catch with next to re-throw the error
// The App will keep loading if you don't use try-catch with next to re-throw the error
app.get("/error-async-await", async (req, res, next) => {
  try {
    const data = await promiseFn();
  } catch (e) {
    // It keeps loading if I don't specify catch.
    next("Catched async/await error with try-catch: " + e);
  }
  console.log(data);
});

app.listen(3000, () => console.log("Example app listening on port 3000!"));
```

---

By the way, you actually pass error to Express when calling `next`.

I feel the documentation is easier to understand after writing a prototype.
http://expressjs.com/en/guide/error-handling.html